js/sri4node-attachments.js

```diff
@@ -1,1078 +1,1677 @@
-      debug('Deleting file ' + remoteFilename);
-      try {
-        await deleteFromS3([remoteFilename])
-        return { status: 204 };
-      } catch (err) {
-        error('Unable to delete file [' + remoteFilename + ']');
-        error(err);
-        throw new sriRequest.SriError({
-          status: 500,
-          errors: [{
-            code: 'delete.failed',
-            type: 'ERROR',
-            message: 'Unable to delete file [' + remoteFilename + ']'
-          }]
-        })
+    // allow everything if no security plugin has been configured
+    if (fullPluginConfig.security && fullPluginConfig.security.plugin) {
+      const security = fullPluginConfig.security.plugin;
+      let attAbility = ability;
+      if (fullPluginConfig.security.abilityPrepend) {
+        attAbility = fullPluginConfig.security.abilityPrepend + attAbility;
+      }
+      if (fullPluginConfig.security.abilityAppend) {
+        attAbility += fullPluginConfig.security.abilityAppend;
       }
-      // }
+      const t = [...resources];
+      await security.checkPermissionOnResourceList(
+        tx,
+        sriRequest,
+        attAbility,
+        t,
+        undefined,
+        true
+      );
     }
+  }
 
+  /**
+   * It checks through all the attachments in the bodyJson if they have a fileHref. If they do,
+   * it will copy that existing file to the current attachment. If the fileHref is not present,
+   * nothing will be done.
+   *
+   * I think this is both used in customRouteForUpload and customRouteForUploadCopy, because
+   * then copies are also supported in the multi-part upload body? But if this is the case,
+   * you could wonder why we also need a separate customRouteForUploadCopy?
+   *
+   * As in so many functions, this function will also return a modified version of the bodyJson.
+   * And it can also throw an SriError, both when something is not right with the data, but also
+   * when the security check fails.
+   *
+   * @param {IDatabase} tx
+   * @param {TSriRequest} sriRequest
+   * @param {Array<TBodyForFileCopy>} toCopy
+   * @param {(href: string) => string} getResourceHrefByAttachmentHref translates the href of the
+   *    attachment to the resource href
+   * @throws {SriError}
+   * @returns {Promise<Array<TMultiPartSingleBodyForFileUploadsWithFileObj & TBodyForFileCopy>>}
+   */
+  async function copyAttachments(
+    tx,
+    sriRequest,
+    toCopy,
+    getResourceHrefByAttachmentHref
+  ) {
+    if (toCopy.length > 0) {
+      sriRequest.logDebug(logChannel, "copy attachments");
+      const resources = new Set(
+        toCopy.map((body) => getResourceHrefByAttachmentHref(body.fileHref))
+      );
+
+      /** @type { Array<TMultiPartSingleBodyForFileUploadsWithFileObj & TBodyForFileCopy> } */
+      const toCopyWithFileObj = toCopy.map((body) => {
+        const resourceHref = getResourceHrefByAttachmentHref(body.fileHref);
+        resources.add(resourceHref);
```

Comment:
The above two lines can be removed I think since the Set has already been initialised.